File: source/yeon/8주차/code/10942팰린드롬.cpp
// 백준 10942 팰린드롬
// cpp solved by dynamic programming
#include<iostream>
#include<vector>
int N, M, s, e;
int arr[2000];
std::vector<std::vector<int>> dp;

int main(void){
    scanf("%d", &N);
    for(int i = 0; i < N; i++){
        scanf("%d", &arr[i]);
    }
    // assign nxn size of vector 
    for(int i = 0; i < N; i++){
        std::vector<int>element(N);
        dp.push_back(element);
    }

    // one element must be the palindrome
    for(int i = 0; i < N; i++) 
        dp[i][i] = 1;

    // two element gonna be palindrome 
    // if first and second element is same
    for(int i = 0; i < N-1; i++){
        if(arr[i]==arr[i+1]) 
            dp[i][i+1] = 1;
    }
    
    // if range is more than 3 
    // just check the first and last element are same
    // and check dp[first+1][last-1] == 1 
    for(int l = 2; l < N; l++){
        for(int i = 0; i < N-l; i++){
            if(arr[i]==arr[i+l] && dp[i+1][i+l-1]==1){
                dp[i][i+l] = 1;
            }
        }
    }
    scanf("%d", &M);
    for(int i = 0; i < M; i++){
        scanf("%d %d", &s, &e);
        printf("%d\n",dp[s-1][e-1]);
    }
}

File: source/yeon/8주차/code/10942팰린드롬.py
# 백준 10942 팰린드롬
# python solved by dynamic programming
import sys

N = int(input())
arr = [int(i) for i in sys.stdin.readline().split()]
dp = [[0 for _ in range(N)] for _ in range(N)]

# if s and e is same it's always be palindrome 
# this is case of just one element
for i in range(N):
    dp[i][i] = 1

# if the range of s and e is 2 
# and if element of s and e are same then it's palindrome
for i in range(N-1):
    if arr[i] == arr[i+1]:
        dp[i][i+1] = 1

# if the range of s and e is over than 2 (3~)
# we can just compare the first and last element and see
# if there's palidrome between the range from dp[][] 
for l in range(2,N):
    for i in range(N-l):
        if arr[i] == arr[i+l] and dp[i+1][i+l-1] == 1 :
            dp[i][i+l] = 1


M = int(input())
for _ in range(M):
    s, e = [int(i) for i in sys.stdin.readline().split()]
    print(dp[s-1][e-1])